[Discover] Add default app state extension and log integration data source profiles

davismcphee opened the pull request:
## Summary

This PR adds a new `getDefaultAppState` extension that allows profiles to set default values for select app state properties, currently `columns` and `rowHeight`. It also adds logs data source sub profiles for the following integrations that consume the `getDefaultAppState` extension (only `columns` are used currently):
- System logs  
- Kubernetes container logs 
- Windows logs
- AWS S3 Logs
- Nginx error logs 
- Nginx access logs
- Apache error logs 

The index patterns and default state for the integrations are hardcoded for the initial implementation, but we should change this later to use an API and state provided by the integrations if we continue this approach.

For testing, you can ingest sample data for the integrations using https://github.com/elastic/kibana-demo-data, but you'll need to reindex the data into correctly named data streams for each:
```
log-system_error -> logs-system.system-test
log-k8s_container -> logs-kubernetes.container_logs-test
log-aws_s3 -> logs-aws.s3access-test
log-nginx_error -> logs-nginx.error-test
log-nqinx -> logs-nginx.access-test
log-apache_error -> logs-apache.error-test

POST /_reindex
{
  "source": {
    "index": "log-k8s_container"
  },
  "dest": {
    "index": "logs-kubernetes.container_logs-test",
    "op_type": "create"
  }
}
```

![default_state](https://github.com/user-attachments/assets/ed73f527-bb5a-470e-b132-f626f0562e18)

Resolves #186271.

### Checklist

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented:
/ci

davismcphee commented on src/plugins/discover/public/application/main/hooks/use_esql_mode.ts:
While working on the necessary changes to this file, I found the code pretty hard to follow. I did some refactoring here that (hopefully) helps clean things up a bit, but I can revert to just the necessary changes if there are concerns with it.

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

davismcphee commented:
/ci

jughosta commented:
Hi @davismcphee,

The code looks great! I have some questions about how it should work.

1. I didn't expect it to keep some of the default columns when switching from  `logs-kubernetes.container_logs-*` to kibana sample logs for example. Is it correct?

![Jul-29-2024 13-33-06](https://github.com/user-attachments/assets/12128179-535f-4145-aea0-e4af8dc7b982)

2. When user presses "New" in the top nav, should it drop or keep the default columns? Currently, it's removing them.

![Jul-29-2024 13-37-30](https://github.com/user-attachments/assets/c2a1be65-0b35-479e-85fd-8f5627e30fe3)

3. The new default columns seem to override `defaultColumns` Advanced Setting. Should not it merge them together? Can we add a functional test for this use case?

davismcphee commented:
@jughosta Thanks for the review!

>I didn't expect it to keep some of the default columns when switching from `logs-kubernetes.container_logs-*` to kibana sample logs for example. Is it correct?

I think so since it works that way when changing data views in general. Do you feel it should work differently here?

>When user presses "New" in the top nav, should it drop or keep the default columns? Currently, it's removing them.

Good question, I kept it like that since I figured it might be useful to have a way to start fresh and remove the defaults, but maybe it's more confusing than helpful. WDYT?

>The new default columns seem to override `defaultColumns` Advanced Setting. Should not it merge them together? Can we add a functional test for this use case?

Also a good question, and more of a product question IMO. I'd prefer not to merge and choose to keep one or the other, but I'm not really sure which to keep.

cc @kertal, might be helpful to get your input on these questions too.

davismcphee commented:
@jughosta As we discussed offline earlier, I updated the PR with these changes:
- Keeping the default columns on switch for now to be consistent with current Discover behaviour, and we can revisit if we get user feedback.
- Reset to default profile state when clicking "New": https://github.com/elastic/kibana/pull/186347/commits/901b24ead05326b33a607e508ac1f549e7b92222
- Merge the default profile columns with the user configured default columns: https://github.com/elastic/kibana/pull/186347/commits/e0ea3daba7aecfad4a6e787eddc1ab3ecf084883

I also added more functional tests including a case for the default columns merging: https://github.com/elastic/kibana/pull/186347/commits/b13c585baae84cc1443df438e165f3cd1173777c.

davismcphee commented:
@jughosta I also noticed a bug when testing where I was waiting until both the records and histogram were finished loading to set the default profile state, which was causing a delay before setting columns in the UI if the records finished fetching before the histogram. I updated the logic to set the default state whenever records are done fetching here: https://github.com/elastic/kibana/pull/186347/commits/a16d651eaa5c0aec29b0f3631dd344aaace038c8. I don't like the callback approach I used to make the update, but I wasn't sure of a better way to do it without significantly refactoring the data fetching.

dmlemeshko reviewed (approved):
`test/functional/services/data_grid.ts` changes LGTM

jughosta reviewed (approved):
LGTM 👍 Thanks for the additional changes and tests!

Something to improve as a follow up maybe is to allow resetting a fixed column width. Otherwise, it looks odd:
![Jul-31-2024 13-48-30](https://github.com/user-attachments/assets/8aa33bf1-3fc4-4127-b268-226ab49a5efb)

For example we could add "Reset width" or "Auto width" to the column menu popover:
<img width="805" alt="Screenshot 2024-07-31 at 13 49 35" src="https://github.com/user-attachments/assets/f02b2be5-bab3-4839-bc73-16f8e479b998">

Wdyt?

logeekal reviewed (approved):
Desk Tested from investigations side. Looks good. Thank you. Left a small comment regarding JSDoc.

davismcphee commented on src/plugins/discover/public/application/main/state_management/discover_app_state_container.ts:
```diff
@@ -57,7 +59,7 @@ export interface DiscoverAppStateContainer extends ReduxLikeStateContainer<Disco
    * @param currentSavedSearch
    */
-  initAndSync: (currentSavedSearch: SavedSearch) => () => void;
+  initAndSync: () => () => void;
```
Good catch! Updated here: https://github.com/elastic/kibana/pull/186347/commits/59c79a8786d2be41d47c54052b7c34bd24455921.

davismcphee commented:
@jughosta I created an issue for the fixed width column problem here: https://github.com/elastic/kibana/issues/189817.
